### ISSUE TITLE
v4.3.0:: compilation error (due to -Werror)

### ISSUE THREAD

**Issue by adriansev** [warmcat/libwebsockets#2493]
Hi! I'm trying to compile v.4.3.0 on fedora 34 `gcc (GCC) 11.2.1 20210728` and i get the following:
```
2021-11-12@17:48:41: [ 21%] Building C object 
lib/CMakeFiles/websockets.dir/tls/openssl/openssl-session.c.o
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/tls.c: In 
function '_lws_tls_restrict_return':
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/tls.c:98:29: 
error: unused variable 'cx' [-Werror=unused-variable]
2021-11-12@17:48:41:    98 |         struct lws_context *cx = 
wsi->a.context;
2021-11-12@17:48:41:       |                             ^~
2021-11-12@17:48:41: cc1: all warnings being treated as errors
2021-11-12@17:48:41: make[2]: *** 
[lib/CMakeFiles/websockets.dir/build.make:188: lib/CMakeFiles/websockets.dir/tls/tls.c.o] Error 1
2021-11-12@17:48:41: make[2]: *** Waiting for unfinished jobs....
2021-11-12@17:48:41: [ 23%] Building C object 
lib/CMakeFiles/websockets.dir/tls/openssl/openssl-client.c.o
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
x509.c: In function 'lws_tls_openssl_cert_info':
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
x509.c:146:30: error: implicit declaration of function 'X509_get_key_usage'; did you mean 'X509_get_ex_data'? 
[-Werror=implicit-function-declaration]
2021-11-12@17:48:41:   146 |                 buf->usage = 
X509_get_key_usage(x509);
2021-11-12@17:48:41:       |                              
^~~~~~~~~~~~~~~~~~
2021-11-12@17:48:41:       |                              X509_get_ex_data
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
x509.c:146:30: error: conversion to 'unsigned int' from 'int' may change the sign of the result 
[-Werror=sign-conversion]
2021-11-12@17:48:41: cc1: all warnings being treated as errors
2021-11-12@17:48:41: make[2]: *** 
[lib/CMakeFiles/websockets.dir/build.make:244: lib/CMakeFiles/websockets.dir/tls/openssl/openssl-x509.c.o] Error 1
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
client.c: In function 'lws_tls_client_create_vhost_context':
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
client.c:738:32: error: implicit declaration of function 'TLS_client_method'; did you mean 'DTLS_client_method'? 
[-Werror=implicit-function-declaration]
2021-11-12@17:48:41:   738 |         method = (SSL_METHOD 
*)TLS_client_method();
2021-11-12@17:48:41:       |                                
^~~~~~~~~~~~~~~~~
2021-11-12@17:48:41:       |                                
DTLS_client_method
2021-11-12@17:48:41: 
/home/physics-tools/alicebuild/sw/BUILD/67b29899477dda23c6a750506ca86fbbc95a81a8/libwebsockets/lib/tls/openssl/openssl-
client.c:738:18: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
2021-11-12@17:48:41:   738 |         method = (SSL_METHOD 
*)TLS_client_method();
2021-11-12@17:48:41:       |                  ^
2021-11-12@17:48:41: cc1: all warnings being treated as errors
```

any idea about these? Is there a way (cmake option) to make it build without applying a patch on the source?
Thanks a lot!!
Adrian

**Comment by lws-team** [comment]
Lws is developed on Fedora... currently on 35 here but that tag was on 34.

Where did your openssl-devel come from and how old is it?

Have you installed more than one openssl or similar library in your filesystem in different places?

**Comment by lws-team** [comment]
... btw it's better to follow v4.3-stable, it has fixes from after v4.3.0 (but not for these).

**Comment by adriansev** [comment]
Hi! openssl is 1.0.2 , and for our (experiment) usage, we cannot build on branches only on tags (to make sure that all the software is the same no matter when was built). Is the el7 openssl no longer supported?
Thanks a lot!

**Comment by lws-team** [comment]
1.0.2 is not supported by openssl themselves any more.

https://www.openssl.org/policies/releasestrat.html

It should build OK on lws, but if you have multiple versions of openssl in your filesystem, the f34 packaged one and this other one, you must take care that lws finds the one you intend and not the distro one, or some mixture.

cmake studies your openssl when it first runs and finds out what apis it supports... the kind of problem you are having looks like, eg, cmake looked at different headers than you are actually compiling, or some other problem with assessing what apis are in there.  Take a look at the second part of this

https://libwebsockets.org/git/libwebsockets/tree/READMEs/README.cmake.md?h=main

**Comment by adriansev** [comment]
We specify the exact openssl that we use, so there should be no mixup, see https://github.com/adriansev/alidist/blob/libwebsockets430/libwebsockets.sh#L32

Is there something lacking? 
Thanks a lot!

**Comment by lws-team** [comment]
Here is an example in CI that directs the build to openssl v3, which is not what the builder has itself.

https://libwebsockets.org/sai/index.html?task=22ac5a9b803dd96862035197d380e76a21719f9b4bbbe58aad7ce7b05f60c4e3

```
cmake .. '-DLWS_OPENSSL_LIBRARIES=/usr/local/src/openssl/v3/usr/local/lib64/libssl.a;/usr/local/src/openssl/v3/usr/local/lib64/libcrypto.a' -DLWS_OPENSSL_INCLUDE_DIRS=/usr/local/src/openssl/v3/usr/local/include/ -DLWS_WITH_MINIMAL_EXAMPLES=1 -DLWS_WITH_GENCRYPTO=0
```

**Comment by adriansev** [comment]
Hi! Thanks for the hint but ...
1. it does not seem to be related to the warnings treated as errors for things like unused variables 
2. using LWS_OPENSSL_LIBRARIES and LWS_OPENSSL_INCLUDE_DIRS it seems that work differently vs. OPENSSL_
with LWS_ options i have 
```
2021-11-12@19:08:54:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: Compiling with SSL support
2021-11-12@19:08:54:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL include dir: /home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/include
2021-11-12@19:08:54:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL libraries: /home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/lib
...
2021-11-12@19:08:56:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: Searching for OpenSSL executable and dlls
2021-11-12@19:08:56:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL executable: /usr/bin/openssl
```
while using the former OPENSSL_ i have :
```
2021-11-12@19:11:43:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: Compiling with SSL support
2021-11-12@19:11:44:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: -- Found OpenSSL: /home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/lib/libcrypto.so (found version "1.0.2o")  
2021-11-12@19:11:44:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL include dir: /home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/include
2021-11-12@19:11:44:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL libraries: /home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/lib/libssl.so;/home/physics-tools/alicebuild/sw/fedora34_x86-64/OpenSSL/v1.0.2o-local1/lib/libcrypto.so;ssl;crypto
...
2021-11-12@19:11:45:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: Searching for OpenSSL executable and dlls
2021-11-12@19:11:45:DEBUG:JAliEn-ROOT:libwebsockets:libwebsocket: OpenSSL executable: /usr/bin/openssl
```
The problem that is common is the wrong executable, and i'm not sure about the libraries part .. 

So, at this point my question would be: do you thing that -Werror breakage comes from the wrong detection of openssl?
Thanks a lot!

**Comment by lws-team** [comment]
Are you sure you need help here :-)  You seem to be telling me how it is and blowing off what I am telling you.

Here is another CI build on main that is on a Centos 7 VM building fine with 1.0.2 openssl.

https://libwebsockets.org/sai/index.html?task=22ac5a9b803dd96862035197d380e76a52a10618aefe6a03698536fa7dd7c273

So I think there is no general problem with lws building on 1.0.2.

**Comment by adriansev** [comment]
So, I was just communicating the observation that on both cases, compilation breaks. The subject of this issues is on the "compilation error" when using "non system provided openssl", i apologize if i was not clear enough.
The context is a new gcc 11.2, external old openssl (but no errors on this at this point), and libwebsockets 4.3.0 tag

Anyway, i think that the discussion got side-tracked: the errors are in the project, and seems to be code related warnings that are converted to errors, and i see no errors related to linking to the external openssl (if there were problems related to usage of openssl i believe that would have take another form).
 
Moreover i reproduced with system ssl on fedora 34 the same error:
```
cmake -DCMAKE_BUILD_TYPE=RELEASE -DLWS_WITHOUT_SERVER=ON -DLWS_WITH_STATIC=ON -DLWS_WITH_SHARED=OFF  -DLWS_WITH_IPV6=ON -DLWS_WITH_ZLIB=OFF -DLWS_HAVE_OPENSSL_ECDH_H=OFF -DLWS_WITHOUT_TESTAPPS=ON ../
...
[ 19%] Building C object lib/CMakeFiles/websockets.dir/tls/openssl/openssl-ssl.c.o
/home/physics-tools/alicebuild/sw/SOURCES/libwebsockets/v4.3.0/v4.3.0/lib/tls/tls.c: In function ‘_lws_tls_restrict_return’:
/home/physics-tools/alicebuild/sw/SOURCES/libwebsockets/v4.3.0/v4.3.0/lib/tls/tls.c:98:29: error: unused variable ‘cx’ [-Werror=unused-variable]
98 |         struct lws_context *cx = wsi->a.context;
|                             ^~
cc1: all warnings being treated as errors
```
So,  given that this is supposed to work, do you have any idea if is there anything wrong in the above cmake commands that would make this error to appear?
Thank you!

**Comment by lws-team** [comment]
So... as I showed above from what is tested in CI, there's none of these problem with lws building Centos 7 openssl 1.0.2.

I repeat, because you blew it off when I said it the first time, you should use v4.3-stable, find out the HEAD hash if you want to pin it; the stable branches keep their history.  I don't want to waste my time on fixed bugs because you want to use a tag.

The thing you're complaining about was solved over a month ago and the fix is on both main and v4.3-stable

1d2df8a8fee4e87fee7831cff1f61ee21e80cd72
